### ISSUE TITLE
Rule Request: Forbid `in` operator and `for...in` on some iterable objects

### ISSUE THREAD

**Issue by sbj42** [palantir/tslint#2607]
### Rule Request

I was converting some code from using plain objects as maps and sets to using the ES6 `Map` and `Set`  classes.  TypeScript handily caught all of the places where the code did things like `map[key] = value` and `delete map[key]`, but it did not catch uses of `key in map` and `for (const key in map) ...`.  This seems like something tslint would be very useful for.

While just applying the rule to Map and Set would be nice, it might be possible to generalize it, either to any [iterable](http://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#iterables) type, or to any type that isn't `any` or [indexable](http://www.typescriptlang.org/docs/handbook/interfaces.html#indexable-types).

- __TSLint version__: 4.5.1
- __TypeScript version__: 2.2.1
- __Running TSLint via__: CLI / VSCode

#### TypeScript code to lint

```ts
const map = new Map<string,string>();
for (const key in map) { // <-- should use `const [key, value] of map` instead
    if (map.hasOwnProperty(key)) { // <-- maybe also warn about this
        /* ... */
    }
}
if ("thing" in map) { // <-- should use `map.has("thing")` instead
    /* ... */
}
const set= new Set<string>();
for (const key in set) { // <-- should use `const key of set` instead
    if (set.hasOwnProperty(key)) { // <-- maybe also warn about this
        /* ... */
    }
}
if ("thing" in set) { // <-- should use `set.has("thing")` instead
    /* ... */
}
```

**Comment by adidahiya** [comment]
we could probably generalize the [`no-for-in-array` rule](https://palantir.github.io/tslint/rules/no-for-in-array/) to do this. we should name it `no-for-in` and deprecate the existing rule

**Comment by sbj42** [comment]
In fact, [these lines in the `no-for-in-array` test case](https://github.com/palantir/tslint/blob/master/test/rules/no-for-in-array/test.ts.lint#L41) seem to agree this is desired.

I'm having a little trouble coming up with an implementation.  At one of those lines, where I think the for-in expression should have type Map, the `Type` obtained has no symbol, flags is 1 ("any").  I'm not sure how to find out that the type is Map, or better yet whether the type has an implementation for the `Symbol.iterator` property.

**Comment by andy-hanson** [comment]
This could use Microsoft/TypeScript#13502 to check for whether the object is iterable.

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
